iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace iox
+{
+namespace popo
+{
+template <typename ChunkDistributorType>
+inline ChunkSender<ChunkDistributorType>::ChunkSender(MemberType_t* const chunkSenderDataPtr) noexcept
+    : ChunkDistributorType(chunkSenderDataPtr)
+{
+}
+
+template <typename ChunkDistributorType>
+inline const typename ChunkSender<ChunkDistributorType>::MemberType_t*
+ChunkSender<ChunkDistributorType>::getMembers() const noexcept
+{
+    return reinterpret_cast<const MemberType_t*>(ChunkDistributorType::getMembers());
+}
+
+template <typename ChunkDistributorType>
+inline typename ChunkSender<ChunkDistributorType>::MemberType_t*
+ChunkSender<ChunkDistributorType>::getMembers() noexcept
+{
+    return reinterpret_cast<MemberType_t*>(ChunkDistributorType::getMembers());
+}
+
+template <typename ChunkDistributorType>
+inline cxx::expected<mepoo::ChunkHeader*, ChunkSenderError>
+ChunkSender<ChunkDistributorType>::allocate(const uint32_t payloadSize) noexcept
+{
+    // use the chunk stored in m_lastChunk if there is one, there is no other owner and the new payload fits in this
+    // chunk
+    if (getMembers()->m_lastChunk && getMembers()->m_lastChunk.hasNoOtherOwners()
+        && getMembers()->m_lastChunk.getChunkHeader()->m_info.m_usedSizeOfChunk
+               >= getMembers()->m_memoryMgr->sizeWithChunkHeaderStruct(payloadSize))
+    {
+        if (getMembers()->m_chunksInUse.insert(getMembers()->m_lastChunk))
+        {
+            getMembers()->m_lastChunk.getChunkHeader()->m_info.m_payloadSize = payloadSize;
+            getMembers()->m_lastChunk.getChunkHeader()->m_info.m_usedSizeOfChunk =
+                getMembers()->m_memoryMgr->sizeWithChunkHeaderStruct(payloadSize);

```

Comment:
small performance optimization: sizeWithChunkHeaderStruct is already called in the if condition. this could be moved out of the if statement

Reply:
I also extended it to be able to reuse the chunk if the payload size increased but it still fits the chunks